docs/development/sync-protocol.md

```diff
@@ -0,0 +1,431 @@
+---
+title: Santa Sync Server Protocol
+parent: Development
+mermaid: true
+---
+
+{% if page.mermaid %}
+  {% include mermaid.html %}
+{% endif %}
+
+# Overview
+
+This documents describes the protocol between Santa and the Sync Server aka the Sync Protocol. Implementors should be able to use this to create their own Sync Servers.
+
+## Background
+
+Santa can be run and configured with a sync server. This allows an admin to
+easily configure and sync rules across a fleet of macOS systems.  In addition to
+distributing rules using a sync server enables an admin to override some local
+configuration options e.g. lockdown mode on both a fleetwide and per host basis.
+
+# The Santa Sync Server Protocol 50,000ft View
+
+The Sync Protocol is an HTTP/JSON Restful protocol. As such it is
+assumed that both the server and client add `Content-Type` headers are set to
+`application/json`. 
+
+The Sync Protocol is client initiated and consists of 4 request-response
+transactions called stages, `preflight`, `eventupload`, `ruledownload` and `postflight`. 
+A sync may consist of all 4 stages, or just the `ruledownload` stage.
```

Comment:
It can also just consist of event upload - when execution is blocked santad will immediately attempt an event upload for that attempt.